packages/buidler-core/src/internal/buidler-evm/provider/node.ts

```diff
@@ -358,6 +360,12 @@ export class BuidlerNode extends EventEmitter {
     await this._saveTransactionAsReceived(tx);
 
     const block = await this._getNextBlockTemplate();
+
+    const timestamp = await this._getNextUsableBlockTimestamp();
```

Comment:
What about moving this logic into `_getNextBlockTemplate`?

This will have two benefits:
- It will decrease code duplication
- It will make the `nextBlockTimestamp` also be used for calls, enabling further control of time there.

Reply:
Note that calls shouldn't reset the next block timestamp.

Reply:
I updated the code and removed some redundant check-and-set `nextBlockTimestamp`. Now, `_getNextUsableBlockTimestamp` is used by the block template only and will have effects with `runCall` as well.
I still keep the function because it would be too big if we flatten it in block template function. This remains a nice separation of the template logic and the time logic.